### ISSUE TITLE
index.as_chat_engine raises "does not support function calling" error

### ISSUE THREAD

**Issue by gavmor** [gpythomas/llamaindex_rag#3]
Got this funny message when using text-gen plugin:
```
  File "/home/user/workspace/other/llamaindex_rag/chat_server.py", line 81, in chat_with_data
    chat_engine = index.as_chat_engine(
                  ^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/workspace/other/llamaindex_rag/venv/lib/python3.11/site-packages/llama_index/core/indices/base.py", line 460, in as_chat_engine
    return AgentRunner.from_llm(
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/workspace/other/llamaindex_rag/venv/lib/python3.11/site-packages/llama_index/core/agent/runner/base.py", line 286, in from_llm
    return OpenAIAgent.from_tools(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/workspace/other/llamaindex_rag/venv/lib/python3.11/site-packages/llama_index/agent/openai/base.py", line 119, in from_tools
    raise ValueError(
ValueError: Model name gpt-3.5-turbo does not support function calling API. 
```

Is the model name wrong? I haven't looked into `from_tools`, so I don't know if it has hard-coded requirements.

These are my text-gen settings:

![image](https://github.com/user-attachments/assets/2b49b5c4-a635-4ddf-9479-55574e4f3cb4)

**Comment by gavmor** [comment]
Looks like we are pulling a "model" string from Obsidian, then passing it down to `OpenRouter()`--which otherwise (when?) defaults to ""microsoft/wizardlm-2-8x22b"--but the returned instance `class OpenRouter(OpenAILike)` doesn't seem to have a truthy `llm.metadata.is_function_calling_model`.

**Comment by gpythomas** [comment]
**TLDR**: I suspect the text-generator plugin is hard-coding the name of OpenAI models in its settings. If you're using `chat_server.py` to access OpenRouter, try typing `openai/gpt-3.5-turbo` in the Model box of the text-gen plugin, and you should be able to access gpt-3.5-turbo.

**Long answer**: When the Flask server is running and forwarding text-generator requests to OpenRouter (or whatever API you've defined in chat_server.py), clicking the refresh icon to the right of models should update the list of models accordingly. You can also test that with `test_get_models()` in `test_API.ipynb`.

For OpenRouter, the model name should be prefixed by the company name, e.g. `openai/gpt-3.5-turbo`. This works fine for all models with `test_get_models()` in `test_API.ipynb` and for non-OpenAI models with the refresh model icon inside text-generator. For some reason, refreshing the model list in text-generator does not refresh the OpenAI model to their proper OpenRouter name. I suspect their name is hard coded in the text-generator interface. If so, I won't be able to do anything about it. Luckily, you can still specify the model name manually as suggested above. Just tested it and it works fine in my installation.
